### PR TITLE
wit-parser: create a top-level document parser.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR refactors `wit-parser` such that there is now a top-level `Document` parser that can represent multiple worlds and interfaces in a single wit document.

The previous internal `Document` type which represented the wit AST has been renamed to simply `Ast`.

The `World::parse_file` method remains unchanged and will fail if there is not exactly one world definition in the file.

A similar helper method `Interface::parse_file` was added back that fails to parse if there is not exactly one interface definition in the file.

With these changes, the hack in the parsing tests to always have a world has been removed.

### CODE REVIEW

**Review by fibonacci1729 — APPROVED** [review]

File: crates/wit-parser/src/ast.rs [hunk 1]
```diff
@@ -11,59 +11,45 @@ mod resolve;
...
-
-    pub fn interfaces(&self) -> Vec<&Interface<'a>> {
-        self.items
-            .iter()
-            .filter_map(|item| match item {
-                DocumentItem::Interface(item) => Some(item),
-                _ => None,
-            })
-            .collect()
+    pub fn worlds(&self) -> impl Iterator<Item = &World<'a>> {
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/wit-parser/src/ast.rs:28]: Then I'd certainly prefer an `items` method that simply returns a borrow of the concrete collection type and hoist the filtering to the caller; I don't like the pattern of collecting to iterate, which is where `worlds` was used prior.

[the remark this answers is not in the document]
Comment (fibonacci1729) [reply crates/wit-parser/src/ast.rs:28]: 💯 I agree.

**Review by alexcrichton — APPROVED** [review]
Thanks!

File: crates/wit-parser/src/ast.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/wit-parser/src/ast.rs:28]: I'm going to punt on this for now and continue returning `impl Iterator` here.

I think it's a (minuscule) improvement over collecting and neither approach fixes the original issue of having the `_` case in the match where (potentially) future AST items might not end up in the iteration.

[the remark this answers is not in the document]
Comment (peterhuene) [reply crates/wit-parser/src/ast.rs:28]: That said, I did replace the `_` case with the (currently) only other case type for each, so at least new cases will point to these iterator methods as problematic in the future.

**Pull request merged**